### ISSUE TITLE
[mimecast] Events missing `Message-ID` header value cause `_id` fingerprint collisions

### ISSUE THREAD

**Issue by ebeahan** [elastic/integrations#9048]
Mimecast data stream ingest pipeline use a `fingerprint` processor to set the `_id` field to prevent duplicates event from indexing. Example config snippet from `siem_logs` data stream pipeline:

```
  - fingerprint:
      fields:
        - mimecast.MsgId
        - mimecast.datetime
      target_field: "_id"
      ignore_missing: true
```

The `mimecast.MsgId` field comes from the `Message-Id` header of the original message ([source](https://integrations.mimecast.com/documentation/tutorials/understanding-siem-logs/)). `Message-Id` values are globally unique, but the field itself isn't required (message `SHOULD` include per https://datatracker.ietf.org/doc/html/rfc2822#section-3.6.4).

Since `ignore_missing: true` is set in the config, the `fingerprint` processor will only have `mimecast.datetime` to generate the fingerprint. This may cause collisions if events contain the same original timestamp. Indexing will fail with 409 errors:

` "status"=>409, "error"=>{"type"=>"version_conflict_engine_exception", "reason"=>"[yL7dnsldqEYTe4cLmr6+wWu/3pU=]: version conflict, document already exists (current version [1])`

**Comment by jamiehynds** [comment]
FYI @mbudge

**Comment by mbudge** [comment]
Check the text file in the support ticket. 

Many integrations are affected.

**Comment by mbudge** [comment]
Here's some of the other integrations I found with 409 version conflicts 

crowdstrike fdr
cisco duo 
o365

We have 100 million 409 version conflict errors in the logstash logs over the last 6 months. This was causing data loss in all the ones logs checked when comparing 409 error logs with logs in elastic with the same ID.

De-duplication is complicated. I wouldn't do it.

**Comment by mbudge** [comment]
Thanks ignore my comments about the other integrations. We have an AWS load balancer causing 409 version conflicts which we'll be fixing soon.

**Comment by mbudge** [comment]
acode is unique per email however there are multiple stages in the email flow through mimecast

receipt
process 
delivery

Each stage is described here https://integrations.mimecast.com/documentation/tutorials/understanding-siem-logs/

We want all 3 because an email might be received, but a policy might block it from being sent to the recipients.

The logs in elastic need to have a field which says receipt, process or delivery so we know what log we are looking at.

On the emails which got rejected they have different acode values Eo7_3S4ROoSY9H01I19v3g  and g1EgOVheNRenXgc12BkbhA

There is normally 1 received and 1 processed log when an email is received, however there can be multiple delivery logs when there are multiple recipients. All these logs have the same acode, so a code can help de-duplicate but extra logic is required. 

I would have 3 fingerprint processors for each log stage, and include all fields from each log type with ignore missing set.

So for a process log include all 17 fields in the fingerprint processor.

**Comment by efd6** [comment]
@mbudge From the doc you linked, t looks to me like we should be able to uniquely distinguish email×stage with the following code in the pipeline.
```
  ### NOTE STAGE
  - set:
      field: mimecast.stage
      value: receipt
      if: ctx.mimecast?.Cphr != null && ctx.mimecast?.IP != null && ctx.mimecast?.Delivered == null
  - set:
      field: mimecast.stage
      value: process
      if: ctx.mimecast?.Cphr == null && ctx.mimecast?.IP == null
  - set:
      field: mimecast.stage
      value: delivery
      if: ctx.mimecast?.Cphr != null && ctx.mimecast?.IP != null && ctx.mimecast?.Delivered != null

  - fingerprint:
      fields:
        - mimecast.MsgId
        - mimecast.aCode
        - mimecast.datetime
        - mimecast.Sender
        - mimecast.Rcpt
        - mimecast.Attempt
        - mimecast.stage
      target_field: "_id"
      ignore_missing: true
```
I don't think we need all fields from each stage since we are just trying to distinguish them (I suspect that not all of the fields that I have there are required). WDYT?

**Comment by mbudge** [comment]
I've added some Go code I used to determine each Mimecast log type to this support ticket #01554878. I used more than 1 field for redundancy in case the field was missing. This code was running for 3-4 years and we didn't spot any problems.

Journal logs are important as they are for internal emails between staff, but journal logs aren't documented on the "Understanding SIEM logs" page. The code I've provided shows how to detect journal logs.

Thanks
